File: native/crates/dodge-batch/src/board.rs
use dodge_core::{FullState, PatternRect, PicoFixed};

pub const BOARD_SIZE: usize = 16;
pub const BOARD_WIDTH: usize = BOARD_SIZE;
pub const BOARD_HEIGHT: usize = BOARD_SIZE;
pub const BOARD_CHANNELS: usize = 19;
pub const BOARD_VALUES: usize = BOARD_CHANNELS * BOARD_HEIGHT * BOARD_WIDTH;
const CELL_VALUES: usize = BOARD_HEIGHT * BOARD_WIDTH;
const VELOCITY_VALUES: usize = 6 * CELL_VALUES;
const SCREEN_SIZE: f32 = 128.0;

pub const PLAYER_PRESENCE: usize = 0;
pub const PLAYER_VX: usize = 1;
pub const PLAYER_VY: usize = 2;
pub const PLAYER_WIDTH: usize = 3;
pub const PLAYER_HEIGHT: usize = 4;
pub const ENEMY_PRESENCE: usize = 5;
pub const ENEMY_VX: usize = 6;
pub const ENEMY_VY: usize = 7;
pub const ENEMY_WIDTH: usize = 8;
pub const ENEMY_HEIGHT: usize = 9;
pub const ENEMY_STAGE: usize = 10;
pub const AOE_PRESENCE: usize = 11;
pub const AOE_VX: usize = 12;
pub const AOE_VY: usize = 13;
pub const AOE_WIDTH: usize = 14;
pub const AOE_HEIGHT: usize = 15;
pub const AOE_STAGE: usize = 16;
pub const AOE_EXPLOSION: usize = 17;
pub const AOE_PATTERN: usize = 18;

/// Channel-major `(19, 16, 16)` semantic board, flattened in C order.
///
/// This is a compatibility observation for the existing CNN. The canonical
/// native observation remains `FullState` plus indexed pixels. Normal enemy
/// records use the same width rule as the Python bridge: personalities 0/1
/// use their current size, while power-up personalities use an 8-pixel box.
#[derive(Clone, Debug, PartialEq)]
pub struct Board19x16 {
    values: Vec<f32>,
}

impl Board19x16 {
    pub fn from_full_state(state: &FullState) -> Self {
        let mut board = vec![0.0; BOARD_VALUES];
        let mut velocity_sums = vec![0.0; VELOCITY_VALUES];
        let mut velocity_counts = vec![0_u32; VELOCITY_VALUES];
        paint_entity(
            &mut board,
            &mut velocity_sums,
            &mut velocity_counts,
            state.player.x,
            state.player.y,
            state.player.size,
            state.player.size,
            state.player.vx,
            state.player.vy,
            None,
            PLAYER_PRESENCE,
            PLAYER_VX,
            PLAYER_VY,
            PLAYER_WIDTH,
            PLAYER_HEIGHT,
            None,
            0,
            None,
        );
        for enemy in &state.enemies {
            if enemy.personality == -1 {
                paint_entity(
                    &mut board,
                    &mut velocity_sums,
                    &mut velocity_counts,
                    enemy.x,
                    enemy.y,
                    enemy.size,
                    enemy.size,
                    enemy.vx,
                    enemy.vy,
                    Some(PicoFixed::ZERO),
                    AOE_PRESENCE,
                    AOE_VX,
                    AOE_VY,
                    AOE_WIDTH,
                    AOE_HEIGHT,
                    Some(AOE_STAGE),
                    4,
                    Some(AOE_EXPLOSION),
                );
            } else {
                let size = if enemy.personality >= 2 {
                    PicoFixed::from_int(8)
                } else {
                    enemy.size
                };
                paint_entity(
                    &mut board,
                    &mut velocity_sums,
                    &mut velocity_counts,
                    enemy.x,
                    enemy.y,
                    size,
                    size,
                    enemy.vx,
                    enemy.vy,
                    Some(PicoFixed::ZERO),
                    ENEMY_PRESENCE,
                    ENEMY_VX,
                    ENEMY_VY,
                    ENEMY_WIDTH,
                    ENEMY_HEIGHT,
                    Some(ENEMY_STAGE),
                    2,
                    None,
                );
            }
        }
        if let Some(pattern_index) = state.active_pattern
            && let Some(pattern) = state.patterns.get(pattern_index)
        {
            for rect in &pattern.rects {
                paint_pattern(&mut board, &mut velocity_sums, &mut velocity_counts, rect);
            }
        }
        write_velocity_channel(&mut board, &velocity_sums, &velocity_counts, PLAYER_VX, 0);
        write_velocity_channel(&mut board, &velocity_sums, &velocity_counts, PLAYER_VY, 1);
        write_velocity_channel(&mut board, &velocity_sums, &velocity_counts, ENEMY_VX, 2);
        write_velocity_channel(&mut board, &velocity_sums, &velocity_counts, ENEMY_VY, 3);
        write_velocity_channel(&mut board, &velocity_sums, &velocity_counts, AOE_VX, 4);
        write_velocity_channel(&mut board, &velocity_sums, &velocity_counts, AOE_VY, 5);
        Self { values: board }
    }

    pub fn as_slice(&self) -> &[f32] {
        &self.values
    }

    /// Return the C-order flat index for a channel-major board coordinate.
    pub const fn flat_index(channel: usize, row: usize, column: usize) -> Option<usize> {
        if channel >= BOARD_CHANNELS || row >= BOARD_HEIGHT || column >= BOARD_WIDTH {
            return None;
        }
        Some(channel * CELL_VALUES + row * BOARD_WIDTH + column)
    }

    /// Read one semantic board cell without exposing unchecked indexing.
    pub fn value(&self, channel: usize, row: usize, column: usize) -> Option<f32> {
        Self::flat_index(channel, row, column).and_then(|index| self.values.get(index).copied())
    }

    pub fn shape(&self) -> (usize, usize, usize) {
        (BOARD_CHANNELS, BOARD_HEIGHT, BOARD_WIDTH)
    }
}

fn paint_pattern(
    board: &mut [f32],
    velocity_sums: &mut [f32],
    velocity_counts: &mut [u32],
    rect: &PatternRect,
) {
    paint_entity(
        board,
        velocity_sums,
        velocity_counts,
        rect.x,
        rect.y,
        rect.width,
        rect.height,
        rect.dx,
        rect.dy,
        Some(rect.sh),
        AOE_PRESENCE,
        AOE_VX,
        AOE_VY,
        AOE_WIDTH,
        AOE_HEIGHT,
        Some(AOE_STAGE),
        4,
        Some(AOE_PATTERN),
    );
}

#[allow(clippy::too_many_arguments)]
fn paint_entity(
    board: &mut [f32],
    velocity_sums: &mut [f32],
    velocity_counts: &mut [u32],
    x: PicoFixed,
    y: PicoFixed,
    width: PicoFixed,
    height: PicoFixed,
    vx: PicoFixed,
    vy: PicoFixed,
    stage: Option<PicoFixed>,
    presence_channel: usize,
    _vx_channel: usize,
    _vy_channel: usize,
    width_channel: usize,
    height_channel: usize,
    stage_channel: Option<usize>,
    velocity_slot: usize,
    kind_channel: Option<usize>,
) {
    let Some((top, bottom, left, right)) = cell_bounds(x, y, width, height) else {
        return;
    };
    let normalized_width = width.to_f32() / SCREEN_SIZE;
    let normalized_height = height.to_f32() / SCREEN_SIZE;
    let normalized_vx = vx.to_f32() / SCREEN_SIZE;
    let normalized_vy = vy.to_f32() / SCREEN_SIZE;
    let normalized_stage = stage.map(|value| value.to_f32() / 2.0);
    for row in top..=bottom {
        for column in left..=right {
            let cell = row * BOARD_WIDTH + column;
            set_presence(board, presence_channel, cell);
            update_max(board, width_channel, cell, normalized_width);
            update_max(board, height_channel, cell, normalized_height);
            add_velocity(
                velocity_sums,
                velocity_counts,
                velocity_slot,
                cell,
                normalized_vx,
                normalized_vy,
            );
            if let (Some(channel), Some(value)) = (stage_channel, normalized_stage) {
                update_max(board, channel, cell, value);
            }
            if let Some(channel) = kind_channel {
                set_presence(board, channel, cell);
            }
        }
    }
}

fn cell_bounds(
    x: PicoFixed,
    y: PicoFixed,
    width: PicoFixed,
    height: PicoFixed,
) -> Option<(usize, usize, usize, usize)> {
    let scale = BOARD_SIZE as f32 / SCREEN_SIZE;
    let left = ((x.to_f32() - width.to_f32() / 2.0) * scale).floor();
    let right = ((x.to_f32() + width.to_f32() / 2.0) * scale).floor();
    let top = ((y.to_f32() - height.to_f32() / 2.0) * scale).floor();
    let bottom = ((y.to_f32() + height.to_f32() / 2.0) * scale).floor();
    let left = left as i32;
    let right = right as i32;
    let top = top as i32;
    let bottom = bottom as i32;
    if right < 0 || left >= BOARD_SIZE as i32 || bottom < 0 || top >= BOARD_SIZE as i32 {
        return None;
    }
    Some((
        top.max(0) as usize,
        bottom.min(BOARD_SIZE as i32 - 1) as usize,
        left.max(0) as usize,
        right.min(BOARD_SIZE as i32 - 1) as usize,
    ))
}

fn set_presence(board: &mut [f32], channel: usize, cell: usize) {
    let index = channel * CELL_VALUES + cell;
    if let Some(value) = board.get_mut(index) {
        *value = 1.0;
    }
}

fn update_max(board: &mut [f32], channel: usize, cell: usize, value: f32) {
    let index = channel * CELL_VALUES + cell;
    if let Some(current) = board.get_mut(index) {
        *current = current.max(value);
    }
}

fn add_velocity(sums: &mut [f32], counts: &mut [u32], slot: usize, cell: usize, vx: f32, vy: f32) {
    let x_index = slot * CELL_VALUES + cell;
    let y_index = (slot + 1) * CELL_VALUES + cell;
    if let Some(sum) = sums.get_mut(x_index) {
        *sum += vx;
    }
    if let Some(sum) = sums.get_mut(y_index) {
        *sum += vy;
    }
    if let Some(count) = counts.get_mut(x_index) {
        *count = count.saturating_add(1);
    }
    if let Some(count) = counts.get_mut(y_index) {
        *count = count.saturating_add(1);
    }
}

fn write_velocity_channel(
    board: &mut [f32],
    sums: &[f32],
    counts: &[u32],
    channel: usize,
    slot: usize,
) {
    for cell in 0..CELL_VALUES {
        let sum_index = slot * CELL_VALUES + cell;
        let Some(count) = counts.get(sum_index).copied() else {
            continue;
        };
        if count == 0 {
            continue;
        }
        let Some(sum) = sums.get(sum_index).copied() else {
            continue;
        };
        let board_index = channel * CELL_VALUES + cell;
        if let Some(value) = board.get_mut(board_index) {
            *value = sum / count as f32;
        }
    }
}

#[cfg(test)]
mod tests {
    use super::{BOARD_CHANNELS, BOARD_HEIGHT, BOARD_SIZE, BOARD_WIDTH, Board19x16};
    use dodge_core::{NativeConfig, NativeGame};

    #[test]
    fn native_ready_state_matches_board_shape_and_finite_contract() {
        let mut game = NativeGame::new(NativeConfig::new(42));
        let mut snapshot = game.reset();
        for _ in 0..13 {
            snapshot = game
                .advance_frame(dodge_core::BUTTON_X_MASK)
                .unwrap_or_else(|_| unreachable!("startup should succeed"))
                .snapshot;
        }
        let board = Board19x16::from_full_state(snapshot.logical_state());
        assert_eq!(board.shape(), (BOARD_CHANNELS, BOARD_HEIGHT, BOARD_WIDTH));
        assert_eq!(
            board.as_slice().len(),
            BOARD_SIZE * BOARD_SIZE * BOARD_CHANNELS
        );
        assert!(board.as_slice().iter().all(|value| value.is_finite()));
        assert_eq!(Board19x16::flat_index(18, 15, 15), Some(4_863));
        assert_eq!(Board19x16::flat_index(19, 0, 0), None);
        assert!(board.value(0, 0, 0).is_some());
    }
}

File: src/dodge/imitation/board.py
from __future__ import annotations

import json
import math
from json import JSONDecodeError
from typing import cast

import numpy as np

from dodge.control import ControlRuntimeError
from dodge.neat.state import (
    SCREEN_SIZE,
    EntityKind,
    EntityState,
    PlayerState,
    RawState,
)

BOARD_SIZE = 16
PLAYER_CHANNEL_NAMES = (
    "player_presence",
    "player_vx",
    "player_vy",
    "player_width",
    "player_height",
)
ENEMY_CHANNEL_NAMES = (
    "enemy_presence",
    "enemy_vx",
    "enemy_vy",
    "enemy_width",
    "enemy_height",
    "enemy_stage",
)
AOE_CHANNEL_NAMES = (
    "aoe_presence",
    "aoe_vx",
    "aoe_vy",
    "aoe_width",
    "aoe_height",
    "aoe_stage",
    "aoe_explosion",
    "aoe_pattern",
)
BOARD_CHANNEL_NAMES = (
    *PLAYER_CHANNEL_NAMES,
    *ENEMY_CHANNEL_NAMES,
    *AOE_CHANNEL_NAMES,
)
BOARD_CHANNELS = len(BOARD_CHANNEL_NAMES)
BOARD_SHAPE = (BOARD_CHANNELS, BOARD_SIZE, BOARD_SIZE)
FULL_BOARD_POSITION_CHANNEL_NAMES = (
    "enemy_x",
    "enemy_y",
    "aoe_x",
    "aoe_y",
)
FULL_BOARD_CHANNEL_NAMES = (*BOARD_CHANNEL_NAMES, *FULL_BOARD_POSITION_CHANNEL_NAMES)
FULL_BOARD_CHANNELS = len(FULL_BOARD_CHANNEL_NAMES)
FULL_BOARD_SHAPE = (FULL_BOARD_CHANNELS, BOARD_SIZE, BOARD_SIZE)

PLAYER_PRESENCE, PLAYER_VX, PLAYER_VY, PLAYER_WIDTH, PLAYER_HEIGHT = range(5)
(
    ENEMY_PRESENCE,
    ENEMY_VX,
    ENEMY_VY,
    ENEMY_WIDTH,
    ENEMY_HEIGHT,
    ENEMY_STAGE,
) = range(5, 11)
(
    AOE_PRESENCE,
    AOE_VX,
    AOE_VY,
    AOE_WIDTH,
    AOE_HEIGHT,
    AOE_STAGE,
    AOE_EXPLOSION,
    AOE_PATTERN,
) = range(11, 19)
ENEMY_X, ENEMY_Y, AOE_X, AOE_Y = range(19, 23)


def raw_state_from_json(encoded: str) -> RawState:
    """Decode the raw state JSON stored beside each learned decision."""
    try:
        payload = _mapping(json.loads(encoded))
        player = _parse_player(payload["player"])
        enemies = _parse_entities(payload["enemies"])
        aoes = _parse_entities(payload["aoes"])
        frame = _integer(payload["frame"])
    except (JSONDecodeError, KeyError, TypeError, ValueError) as error:
        raise ControlRuntimeError("collector raw state JSON is invalid") from error
    return RawState(frame, player, enemies, aoes)


def encode_board(
    state: RawState,
    *,
    include_offscreen: bool = False,
    preserve_coordinates: bool = False,
) -> np.ndarray:
    """Rasterize the complete raw board state into CNN-friendly channels."""
    if preserve_coordinates and not include_offscreen:
        raise ValueError("coordinate preservation requires offscreen entities")
    board = np.zeros(
        FULL_BOARD_SHAPE if preserve_coordinates else BOARD_SHAPE,
        dtype=np.float32,
    )
    velocity_sums = np.zeros((6, BOARD_SIZE, BOARD_SIZE), dtype=np.float32)
    velocity_counts = np.zeros_like(velocity_sums)

    _paint_entity(
        board,
        velocity_sums,
        velocity_counts,
        x=state.player.x,
        y=state.player.y,
        width=state.player.size,
        height=state.player.size,
        vx=state.player.vx,
        vy=state.player.vy,
        stage=None,
        kind_channel=None,
        presence_channel=PLAYER_PRESENCE,
        vx_channel=PLAYER_VX,
        vy_channel=PLAYER_VY,
        width_channel=PLAYER_WIDTH,
        height_channel=PLAYER_HEIGHT,
        stage_channel=None,
        velocity_slot=0,
        include_offscreen=include_offscreen,
        position_channels=None,
    )
    for entity in state.enemies:
        _paint_entity(
            board,
            velocity_sums,
            velocity_counts,
            x=entity.x,
            y=entity.y,
            width=entity.width,
            height=entity.height,
            vx=entity.vx,
            vy=entity.vy,
            stage=entity.stage,
            kind_channel=None,
            presence_channel=ENEMY_PRESENCE,
            vx_channel=ENEMY_VX,
            vy_channel=ENEMY_VY,
            width_channel=ENEMY_WIDTH,
            height_channel=ENEMY_HEIGHT,
            stage_channel=ENEMY_STAGE,
            velocity_slot=2,
            include_offscreen=include_offscreen,
            position_channels=(ENEMY_X, ENEMY_Y) if preserve_coordinates else None,
        )
    for entity in state.aoes:
        _paint_entity(
            board,
            velocity_sums,
            velocity_counts,
            x=entity.x,
            y=entity.y,
            width=entity.width,
            height=entity.height,
            vx=entity.vx,
            vy=entity.vy,
            stage=entity.stage,
            kind_channel=(AOE_EXPLOSION if entity.kind == "explosion" else AOE_PATTERN),
            presence_channel=AOE_PRESENCE,
            vx_channel=AOE_VX,
            vy_channel=AOE_VY,
            width_channel=AOE_WIDTH,
            height_channel=AOE_HEIGHT,
            stage_channel=AOE_STAGE,
            velocity_slot=4,
            include_offscreen=include_offscreen,
            position_channels=(AOE_X, AOE_Y) if preserve_coordinates else None,
        )

    for channel, slot in (
        (PLAYER_VX, 0),
        (PLAYER_VY, 1),
        (ENEMY_VX, 2),
        (ENEMY_VY, 3),
        (AOE_VX, 4),
        (AOE_VY, 5),
    ):
        np.divide(
            velocity_sums[slot],
            velocity_counts[slot],
            out=board[channel],
            where=velocity_counts[slot] > 0,
        )
    return board


def _paint_entity(
    board: np.ndarray,
    velocity_sums: np.ndarray,
    velocity_counts: np.ndarray,
    *,
    x: float,
    y: float,
    width: float,
    height: float,
    vx: float,
    vy: float,
    stage: float | None,
    kind_channel: int | None,
    presence_channel: int,
    vx_channel: int,
    vy_channel: int,
    width_channel: int,
    height_channel: int,
    stage_channel: int | None,
    velocity_slot: int,
    include_offscreen: bool,
    position_channels: tuple[int, int] | None,
) -> None:
    bounds = _cell_bounds(x, y, width, height, include_offscreen=include_offscreen)
    if bounds is None:
        return
    top, bottom, left, right = bounds
    normalized_width = width / SCREEN_SIZE
    normalized_height = height / SCREEN_SIZE
    normalized_vx = vx / SCREEN_SIZE
    normalized_vy = vy / SCREEN_SIZE
    normalized_stage = None if stage is None else stage / 2
    normalized_x = 0.5 + x / SCREEN_SIZE
    normalized_y = 0.5 + y / SCREEN_SIZE
    for row in range(top, bottom + 1):
        for column in range(left, right + 1):
            board[presence_channel, row, column] = 1.0
            board[width_channel, row, column] = max(
                board[width_channel, row, column], normalized_width
            )
            board[height_channel, row, column] = max(
                board[height_channel, row, column], normalized_height
            )
            velocity_sums[velocity_slot, row, column] += normalized_vx
            velocity_sums[velocity_slot + 1, row, column] += normalized_vy
            velocity_counts[velocity_slot, row, column] += 1
            velocity_counts[velocity_slot + 1, row, column] += 1
            if stage_channel is not None and normalized_stage is not None:
                board[stage_channel, row, column] = max(
                    board[stage_channel, row, column], normalized_stage
                )
            if kind_channel is not None:
                board[kind_channel, row, column] = 1.0
            if position_channels is not None:
                x_channel, y_channel = position_channels
                board[x_channel, row, column] = max(
                    board[x_channel, row, column], normalized_x
                )
                board[y_channel, row, column] = max(
                    board[y_channel, row, column], normalized_y
                )


def _cell_bounds(
    x: float,
    y: float,
    width: float,
    height: float,
    *,
    include_offscreen: bool = False,
) -> tuple[int, int, int, int] | None:
    scale = BOARD_SIZE / SCREEN_SIZE
    left = math.floor((x - width / 2) * scale)
    right = math.floor((x + width / 2) * scale)
    top = math.floor((y - height / 2) * scale)
    bottom = math.floor((y + height / 2) * scale)
    if not include_offscreen and (
        right < 0 or left >= BOARD_SIZE or bottom < 0 or top >= BOARD_SIZE
    ):
        return None
    if right < 0:
        left = right = 0
    elif left >= BOARD_SIZE:
        left = right = BOARD_SIZE - 1
    else:
        left = max(0, left)
        right = min(BOARD_SIZE - 1, right)
    if bottom < 0:
        top = bottom = 0
    elif top >= BOARD_SIZE:
        top = bottom = BOARD_SIZE - 1
    else:
        top = max(0, top)
        bottom = min(BOARD_SIZE - 1, bottom)
    return (
        top,
        bottom,
        left,
        right,
    )


def _parse_player(value: object) -> PlayerState:
    payload = _mapping(value)
    return PlayerState(
        _number(payload["x"]),
        _number(payload["y"]),
        _number(payload["vx"]),
        _number(payload["vy"]),
        _number(payload["size"]),
    )


def _parse_entities(value: object) -> tuple[EntityState, ...]:
    return tuple(_parse_entity(item) for item in _sequence(value))


def _parse_entity(value: object) -> EntityState:
    payload = _mapping(value)
    kind_value = payload["kind"]
    if kind_value not in {"enemy", "explosion", "pattern"}:
        raise ValueError("unknown entity kind")
    return EntityState(
        _number(payload["x"]),
        _number(payload["y"]),
        _number(payload["vx"]),
        _number(payload["vy"]),
        _number(payload["width"]),
        _number(payload["height"]),
        cast(EntityKind, kind_value),
        _number(payload["stage"]),
    )


def _mapping(value: object) -> dict[str, object]:
    if not isinstance(value, dict):
        raise ValueError("expected JSON object")
    return value


def _sequence(value: object) -> list[object]:
    if not isinstance(value, list):
        raise ValueError("expected JSON list")
    return value


def _number(value: object) -> float:
    if isinstance(value, bool) or not isinstance(value, int | float):
        raise ValueError("expected JSON number")
    return float(value)


def _integer(value: object) -> int:
    if isinstance(value, bool) or not isinstance(value, int):
        raise ValueError("expected JSON integer")
    return value
